📖 [GEP-22] Improved Usage of the `ShootState` API (#8012)
gardener/gardener: docs/proposals/22-eliminating-shootstate-api.md

---
title: Eliminating the `ShootState` API
gep-number: 22
creation-date: 2023-06-01
status: implementable
authors:
- "@rfranzke"
reviewers:
- "@timebertt"
---

# GEP-22: Eliminating the `ShootState` API

## Table of Contents

<!-- TOC -->
- [GEP-22: Eliminating the `ShootState` API](#gep-22-eliminating-the-shootstate-api)
  - [Table of Contents](#table-of-contents)
  - [Summary](#summary)
  - [Motivation](#motivation)
    - [Goals](#goals)
    - [Non-Goals](#non-goals)
  - [Proposal](#proposal)
  - [Alternatives](#alternatives)
  - [Future Improvements](#future-improvements)
<!-- TOC -->

## Summary

> For each `Shoot` resource, `gardenlet` creates a corresponding `ShootState` resource which has the same lifecycle of the `Shoot`.

Reviewer: ```suggestion
For each `Shoot` resource, `gardenlet` creates a corresponding `ShootState` resource which has the same lifecycle as the `Shoot`.
```

Especially due to the second purpose, the `ShootState` API imposes several scalability concerns and limitations for large Gardener installations.
Hence, this GEP proposes the elimination of said `ShootState` API by replacing it with lightweight alternatives.

This GEP focuses on the second use-case only and ignores the first use-case.
An alternative for it is worked on separately in [Introduce `core.gardener.cloud/v1beta1.InternalSecret` API #7999](https://github.com/gardener/gardener/issues/7999).

## Motivation

The `gardenlet` runs [two controllers](../concepts/gardenlet.md#shootstate-controller) for maintaining the most crucial data of shoot clusters in their `ShootState` resources:

- The [`shootstate-secret`](../../pkg/gardenlet/controller/shootstate/secret) controller watches `Secret`s labeled with `persist=true` in the shoot namespaces in the seed cluster. It persists all labels and data of such `Secret`s into the related `ShootState` resource.

In reality, those `Secret`s only change when the end-user initiates a credentials rotation operation, so rather rarely.

> Still, duplicating all labels and data of such `Secret`s into the garden cluster increases the size of `ShootState` resources.

Reviewer 2: Nit: indent this paragraph

- The [`shootstate-extensions`](../../pkg/gardenlet/controller/shootstate/extensions) controller watches all resources of the `extensions.gardener.cloud/v1alpha1` API group in the shoot namespaces in the seed cluster. It persists all extensions state and resources in the respective `.status.state` and `.status.resources` fields. See also [this document](../extensions/migration.md) for more background information.

In reality, this is the controller which can cause quite a lot of network I/O (hence, traffic and costs).
While most of the extension states change rather rarely, the state of the `extensions.gardener.cloud/v1alpha1.Worker` resources can update quite frequently whenever new worker machines join or leave the cluster.

> `gardenlet` has to watch and replicate all such state changes to the `ShootState` resource.

Reviewer 2: Nit: indent this paragraph

We are proposing the following changes:

1. The [`shootstate-extensions`](../../pkg/gardenlet/controller/shootstate/extensions) controller of `gardenlet` will be dropped, i.e., it will no longer replicate the extensions states to the `ShootState` resources in the garden cluster.
2. The [`shootstate-secret`](../../pkg/gardenlet/controller/shootstate/secret) controller will also be dropped as soon as [Introduce `core.gardener.cloud/v1beta1.InternalSecret` API #7999](https://github.com/gardener/gardener/issues/7999) has been implemented.

With this, the `ShootState` API can be dropped entirely.

3. During the `Migrate` phase of a `Shoot`, `gardenlet` will fetch all relevant state in the shoot namespaces of the current seed, similar to what the two controllers are doing as of today:

   1. all labels and data for `Secret`s labeled with `persist=true`.
   2. all `.status.state` and `.status.resources` for resources in the `extensions.gardener.cloud/v1alpha1` API group.

>    The `gardenlet` puts this state into a structure similar to the [`core.gardener.cloud/v1beta1.ShootSpec`](https://github.com/gardener/gardener/blob/27b76d8d825461b9edc75c2d4472829e309af05e/pkg/apis/core/v1beta1/types_shootstate.go#L49-L101), marshals it to JSON, and puts it into the data of an [`core.gardener.cloud/v1beta1.InternalSecret`](https://github.com/gardener/gardener/issues/7999) stored in the namespace of the respective `Shoot` in the garden cluster.

Reviewer 2: ```suggestion
   The `gardenlet` puts this state into a structure similar to the [`core.gardener.cloud/v1beta1.ShootStateSpec`](https://github.com/gardener/gardener/blob/27b76d8d825461b9edc75c2d4472829e309af05e/pkg/apis/core/v1beta1/types_shootstate.go#L49-L101), marshals it to JSON, and puts it into the data of an [`core.gardener.cloud/v1beta1.InternalSecret`](https://github.com/gardener/gardener/issues/7999) stored in the namespace of the respective `Shoot` in the garden cluster.
```

> 4. During the `Restore` phase of a `Shoot`, `gardenlet` will read the created [`core.gardener.cloud/v1beta1.InternalSecret`](https://github.com/gardener/gardener/issues/7999) from the namespace of the respective `Shoot` in the garden cluster, unmarshals its data from JSON to a structure similar to the [`core.gardener.cloud/v1beta1.ShootSpec`](https://github.com/gardener/gardener/blob/27b76d8d825461b9edc75c2d4472829e309af05e/pkg/apis/core/v1beta1/types_shootstate.go#L49-L101), and uses it to populate the state in the new seed cluster just like today.

Reviewer 2: ```suggestion
4. During the `Restore` phase of a `Shoot`, `gardenlet` will read the created [`core.gardener.cloud/v1beta1.InternalSecret`](https://github.com/gardener/gardener/issues/7999) from the namespace of the respective `Shoot` in the garden cluster, unmarshals its data from JSON to a structure similar to the [`core.gardener.cloud/v1beta1.ShootStateSpec`](https://github.com/gardener/gardener/blob/27b76d8d825461b9edc75c2d4472829e309af05e/pkg/apis/core/v1beta1/types_shootstate.go#L49-L101), and uses it to populate the state in the new seed cluster just like today.
```

  ```yaml
  apiVersion: extensions.gardener.cloud/v1alpha1
  kind: BackupDownload
  metadata:
    name: example
    namespace: default
  spec:
    entryName: example-entry
    filePath: shootstate.yaml.enc
    type: local
  status:
    data: <downloaded-data>
  ```
  </details>

Similar to all other resources in the `extensions.gardener.cloud/v1alpha1` API, we could provide respective controllers in the [extensions library](https://github.com/gardener/gardener/tree/master/extensions) such that provider extensions only need to implement their specific business logic (i.e., uploading/downloading to/from their infrastructure-specific object storage).

However, this approach is significantly more expensive in terms of development costs and complexity since the following additional adaptations would be required:

- Introduction of new resources in the `extensions.gardener.cloud/v1alpha1` API which comes with validation, deployment dependencies, etc.
- Introduction of new extension controllers.
- Adaptation of all existing provider extensions with support for `Backup{Bucket,Entry}`s with the implementation for the new `Backup{Up,Down}load` resources.

> - Encryption of the shoot state data (since we do not want to store it unencryptedly in the backup buckets).

Reviewer 2: ```suggestion
- Encryption of the shoot state data (since we do not want to store it as plain text in the backup buckets).
```

- Management (incl. rotation) of a respective encryption key.

So far, the only justification for following such a path would be the necessity for regular backups of the shoot states (e.g., every hour or so).
In this case, the shoot states would again be stored in the garden cluster for the entire lifetime of the `Shoot`s.

This alternative approach would still improve a few things like network traffic, etc. since the data would no longer be continuously replicated (only with a much lower frequency compared to today), but eventually today's technical debts remain limiting the garden cluster's scalability.
As no appropriate reasoning for regular backups has been presented yet, they were added as "non-goal" in this GEP.
Hence, implementing this alternative approach is considered unnecessary.

## Future Improvements

> With [Move `machine-controller-manager` reconciliation responsibility from extensions to `gardenlet` #7594](https://github.com/gardener/gardener/issues/7594), `gardenlet` will become known of the `machine.sapcloud.io/v1alpha1` API used for managing the worker machines of shoot clusters.

Reviewer: ```suggestion
With [Move `machine-controller-manager` reconciliation responsibility from extensions to `gardenlet` #7594](https://github.com/gardener/gardener/issues/7594), `gardenlet` will become aware of the `machine.sapcloud.io/v1alpha1` API used for managing the worker machines of shoot clusters.
```

This allows to drop the [`Worker` state reconciler](https://github.com/gardener/gardener/blob/master/extensions/pkg/controller/worker/state_reconciler.go) which currently watches all `MachineDeployment`s, `MachineSet`s, and `Machine`s, and replicates their specifications into the `.status.state` of the related `extensions.gardener.cloud/v1alpha1.Worker` resource.

> Instead, `gardenlet` could now collect these resources during the `Migrate` phase itself, effectively compute the necessary `Worker` state, and also restore it during the `Restore` phase.

Reviewer: ```suggestion
Instead, `gardenlet` could then collect these resources during the `Migrate` phase itself, effectively compute the necessary `Worker` state, and also restore it during the `Restore` phase.
```
